src/garbage_collector.rs

```diff
@@ -343,19 +404,86 @@ impl GarbageCollectorOrchestrator {
+        "Compacted SSTs"
+    }
+}
+
+impl GarbageCollectorOrchestrator {
+    async fn load_stored_manifest(&self) -> Result<StoredManifest, SlateDBError> {
+        StoredManifest::load(Arc::clone(&self.manifest_store)).await
+    }
+
+    fn filter_expired_checkpoints(
+        manifest: &StoredManifest,
+    ) -> Result<Option<CoreDbState>, SlateDBError> {
+        let utc_now = Utc::now();
+        let mut db_state = manifest.db_state().clone();
+        let retained_checkpoints: Vec<Checkpoint> = db_state
+            .checkpoints
+            .iter()
+            .filter(|checkpoint| match checkpoint.expire_time {
+                Some(expire_time) => DateTime::<Utc>::from(expire_time) > utc_now,
+                None => true,
+            })
+            .cloned()
+            .collect();
+
+        let updated_state = if db_state.checkpoints.len() != retained_checkpoints.len() {
+            db_state.checkpoints = retained_checkpoints;
+            Some(db_state)
+        } else {
+            None
+        };
+        Ok(updated_state)
+    }
+
+    async fn remove_expired_checkpoints(&self) -> Result<(), SlateDBError> {
+        let mut stored_manifest = self.load_stored_manifest().await?;
+        stored_manifest
+            .maybe_apply_db_state_update(Self::filter_expired_checkpoints)
+            .await
+    }
+
     /// Run the garbage collector
     pub async fn run(&self) {
         let log_ticker = crossbeam_channel::tick(Duration::from_secs(60));
-        let (manifest_ticker, mut manifest_status) =
-            Self::options_to_ticker(self.options.manifest_options.as_ref());
-        let (wal_ticker, mut wal_status) =
-            Self::options_to_ticker(self.options.wal_options.as_ref());
-        let (compacted_ticker, mut compacted_status) =
-            Self::options_to_ticker(self.options.compacted_options.as_ref());
+
+        let mut wal_gc_task = WalGcTask::new(
+            self.manifest_store.clone(),
+            self.table_store.clone(),
+            self.db_stats.clone(),
+            self.options.wal_options,
+        );
+        let mut compacted_gc_task = CompactedGcTask::new(
+            self.manifest_store.clone(),
+            self.table_store.clone(),
+            self.db_stats.clone(),
+            self.options.compacted_options,
+        );
+        let mut manifest_gc_task = ManifestGcTask::new(
+            self.manifest_store.clone(),
+            self.db_stats.clone(),
+            self.options.manifest_options,
+        );
+
+        let manifest_ticker = manifest_gc_task.ticker();
+        let wal_ticker = wal_gc_task.ticker();
+        let compacted_ticker = compacted_gc_task.ticker();
 
         info!(
             "Starting Garbage Collector with [manifest: {}], [wal: {}], [compacted: {}]",
-            manifest_status, wal_status, compacted_status
+            manifest_gc_task.status(),
+            wal_gc_task.status(),
+            compacted_gc_task.status()
         );
 
         loop {
```

Comment:
not part of your patch, but stood out since we changed the select! conditions - shouldn't we be using `tokio::select!` here? `crossbeam_channel::select!` is synchronous so this would block the tokio runtime thread.

Reply:
Below is the code that spwans the thread. Wouldn't spawning a new thread not block the tokio runtime thread? Not sure how all this works in rust. I would imagine the `tokio_handle.block_on` part would handle the synchronicity part, but I'm a bit beyond my depth here. Perhaps we should leave changing this for another PR, since that would also involve changing all the ticker logic.

```rs
        thread::spawn(move || {
            let orchestrator = GarbageCollectorOrchestrator {
                manifest_store,
                table_store,
                options,
                external_rx,
                db_stats,
            };
            tokio_handle.block_on(orchestrator.run());

            // !important: make sure that this is always the last thing that this
            // thread does, otherwise we risk notifying waiters and leaving this
            // thread around after shutdown
            if shutdown_tx.send(true).is_err() {
                error!("Could not send shutdown signal to threads blocked on await_shutdown");
            }
        });
```

Reply:
OK ChatGPT confirms your suspicion, the tokio thread will be blocked on this. I'm not sure why we have `tokio_handle.block_on` if we don't want that?

Reply:
what this is doing is spawning a thread, and having that thread block until `orchestrator.run` completes on the tokio runtime. `orchestrator.run` still runs on the tokio runtime, and so if it blocks then it will block the runtime's threads, which is a no-no. I don't mind switching it over to tokio select after we merge this.

Reply:
https://github.com/slatedb/slatedb/issues/380